### ISSUE TITLE
Windows Authentication: Different Groups returned for Auto-Login or Explicit Manual Login

### ISSUE THREAD

**Issue by RickStrahl** [aspnet/AspNetCore#14766]
I have an ASP.NET Core 3.0 application that works with local Intranet Windows Authentication to identify logged in users. Using the standard Windows Authentication behaviors I'm able to capture the user's `WindowsIdentity` without an issue.

However, depending on how the user is logged into the browser using either automatic Intranet Browser login (ie. no password dialog) or explicitly logging in using the browser Password dialog box, **I get different results for the user's groups**.

The following is an API request that echos back user information including a filtered group membership list (that excludes built-in accounts). The one on the left is a manual login, the one on the right an auto-login.

For the explicit login I correctly see all the custom groups the user is part of. However, for the auto-login, those same groups do not show up:

![image](https://user-images.githubusercontent.com/1374013/66263517-b389c180-e78f-11e9-80b6-7b782d18f077.png)

I also took a close look at the User and Identity instances on the server, and it's referencing the exact same SIDs for the user, so it seems strange that different results are being returned for the Group Membership.

Any ideas why the group list is different when I am getting the same account returned? Note the groups are local so it shouldn't be an issue due to domain access.

> Note: I'm testing locally on localhost even, and to test this I set the Windows Proxy Settings here:
>
> ![image](https://user-images.githubusercontent.com/1374013/66263518-b97fa280-e78f-11e9-97b4-afb8c02d75ea.png)
> 
> With the checkboxes off I'm forced to login. With them on (in Chromium browsers anyway) I have to explicitly enter my credentials into the browser's login dialog.

### Environment
* Windows 10
* Local Logons
* Using Chrome/Edgium (they auto-logon by default)
* Using Kestrel or IIS Express (tried both same behavior)

### Repro Steps
* Create a Web Application with Windows Auth enabled.
* Add an endpoint or middleware Use handler after authentication
* Make sure you have or add some user defined roles (Local is what I'm using here)
* Access via Edge Chrome on local machine  
Expected: Auto-login and you won't be able to see the custom groups
* Now switch to forced logins in Internet Settings
    * Open Internet Settings | Security | Local Intranet Zone
    * Open Advanced | Custom Level | User Authentication | Logon
    * Set to: Prompt for username and password
* Shut down browser and re-run same request    
Expected: you should now properly get the custom groups returned

```csharp
// in `Configure()`
app.UseAuthentication(); 
app.UseAuthorization();

// Pick up Windows Authentication
app.Use(async (context, next) =>
{
    
    // First off ensure that user is authenticated with NTLM Auth
    if (!context.User.Identity.IsAuthenticated)
    {
        context.Response.StatusCode = 401;
        context.Response.Headers.Add("www-authenticate",
            new Microsoft.Extensions.Primitives.StringValues(new string[]
            {
                "Negotiate", "NTLM"
            }));

        await context.Response.WriteAsync("Unauthorized");
        return;
    }

    // check for your group here
    var isInGroup = context.User.IsInRole("FTDB-Technician");
    
    var groups = GetGroups(user.Identity as WindowsIdentity);
    await next();
    
});

// windows identity 
public static List<WindowsGroup> GetGroups(WindowsIdentity identity)
{
    var groups = new List<WindowsGroup>();

    foreach (var group in identity.Groups)
    {
        var sid = group.Value;
        var groupName = GetGroupNameBySid(sid);
        var (username, domain) = SplitUserDomain(groupName);
        
        if (groupName.StartsWith("NT AUTHORITY") ||
            groupName.StartsWith("BUILTIN") ||
            username == "Everyone" || 
            username == "None" ||
            username == "docker-users")
                
            continue;

        groups.Add(new WindowsGroup() {Name = username , Domain = domain + "," + sid + "," + identity.Name});
    }

    return groups;
}

public static string GetGroupNameBySid(string sid)
{
    try
    {
        return new SecurityIdentifier(sid).Translate(typeof(NTAccount)).ToString();
    }
    catch
    {
        return sid;
    }
}
```

### Expected Behavior
Expecting to see the same user data regardless of whether the login is 'automatic' or manual.

**Comment by blowdart** [comment]
Does IsInRole still work for the "missing" groups? WindowsIdentity doesn't populate the roles in the way cookies or JWT does.

**Comment by RickStrahl** [comment]
This is on Windows. Same result using Kestrel directly or using IIS Express.

Again - the problem isn't that it doesn't it work at all - it's that it doesn't work when automatically logged in vs. explicitly being logged. 

Different behavior for what should result in the same account being returned with the same exact attributes. 

The questions that come to my mind is this:

* Is Chrome sending a different user context that makes this different   
depending on whether it uses explicitly forced log in or auto-login?

* Why would ASP.NET Care about above? Windows is picking up the same  
exact user SID, so why the different group memberships here?

Just for reference I'm pasting the request traces for each of the three types of requests I tend to make:

* Http Client (explicitly provides uid/pwd on each request) - works
* Edgium/Chrome with auto-login - doesn't work
* Edgium/Chrome with auto-login disable forcing explicit login - works


Comparing request traces:

**Works (from my HTTP Client):**

```http
GET https://localhost:5001/account/userstate HTTP/1.1
Pragma: no-cache
Cache-Control: no-cache
Upgrade-Insecure-Requests: 1
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3
Sec-Fetch-Site: none
Sec-Fetch-Mode: navigate
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3915.0 Safari/537.36 Edg/79.0.287.3
Authorization: Negotiate TlR22VNTUAADAAAAGAAYAHIAAADuAO4AigAAAAAAAABYAAAADgAOAFgAAAAMAAwAZgAAABAAEAB4AQAAFYKI4goA6kkAAAAPOU/w7N9731GENKwO8PjeZ3IAcwB0AHIAYQBoAGwAUgBBAFMAVwBJAE4AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA9G0vfxZ+ZEoaKItYU6VcUwEBAAAAAAAAGUz1Unt81QHDEL7Hd3KNbAAAAAACAAwAUgBBAFMAVwBJAE4AAQAMAFIAQQBTAFcASQBOAAQADABSAEEAUwBXAEkATgADAAwAUgBBAFMAVwBJAE4ABwAIABlM9VJ7fNUBBgAEAAIAAAAIADAAMAAAAAAAAAABAAAAACAAADmayz0EVIkTD7fPcH/qMN9F2ruH6zvDQe/tOeG8atdtCgAQAKBzSKrJJNkmrnS0V0gR6uIJABYASABUAFQAUAAvAFIAQQBTAFcASQBOAAAAAAAAAAAAAAAAADIZrqxQdxpVnwAAFnZCHMQ=
Host: localhost:5001
```

Doesn't work from Web Browser with Auto-Login:

```http
GET https://localhost:5001/account/userstate HTTP/1.1
Host: localhost:5001
Connection: keep-alive
Pragma: no-cache
Cache-Control: no-cache
Authorization: Negotiate oXcwdaADCgEBoloEWE5UTE1TU1AAAwAAAAAAAABYAAAAAAAAAFgAAAAAAAAAWAAAAAAAAABYAAAAAAAAAFgAAAAAAAAAWAAAABXCiOIKAOpJAAAAD90Uv02K71nzNItLOLlDRtijEgQQAQAAABTPl7JCSaeOAAAAAA==
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3915.0 Safari/537.36 Edg/79.0.287.3
Sec-Fetch-User: ?1
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3
Sec-Fetch-Site: none
Sec-Fetch-Mode: navigate
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
Cookie: __utmz=111872281.1557000280.1.1.utmcsr=(direct)|utmccn=(direct)|utmcmd=(none); WebStoreUser=; _ga=GA1.1.1593582071.1557000280; .AspNet.Consent=yes; weblog_user=Rick Strahl; weblog_email=rstrahl@west-wind.com; wwt=6EYWGHcnCkBkUyCL; __utma=111872281.1593582071.1557000280.1569911705.1569917482.17; ftdb-token=023AD39E6344ACF2DF291525080B9B1AE79823956E5C1FC3351DE8AC24E4CA328F7868AC8F7DE6A8AE8935D4A0C3394D46D6A0CBAFBE597E75008462DF3386E0
```

**Works: In Chrome with Explicit Login Forced via Connection Settings**

```http
GET https://localhost:5001/account/userstate HTTP/1.1
Host: localhost:5001
Connection: keep-alive
Cache-Control: max-age=0
Authorization: Negotiate TlRMTVNTUAADAAAAGAAYAHIAAAD0APQAigAAAAAAAABYAAAADgAOAFgAAAAMAAwAZgAAABAAEAB+AQAAFYKI4goA6kkAAAAPa8DwLNe4FougDeH7X6mJxnIAcwB0AHIAYQBoAGwAUgBBAFMAVwBJAE4AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAi0KZu2oJYAS5lGjLTOvqtwEBAAAAAAAAeHZyDH181QGOd5Wg5/pHlwAAAAACAAwAUgBBAFMAVwBJAE4AAQAMAFIAQQBTAFcASQBOAAQADABSAEEAUwBXAEkATgADAAwAUgBBAFMAVwBJAE4ABwAIAHh2cgx9fNUBBgAEAAIAAAAIADAAMAAAAAAAAAABAAAAACAAADmayz0EVIkTD7fPcH/qMN9F2ruH6zvDQe/tOeG8atdtCgAQAE+uI73nPkm2PKAT+XvLeuMJABwASABUAFQAUAAvAGwAbwBjAGEAbABoAG8AcwB0AAAAAAAAAAAAAAAAAEYojwOzCSd/O7DIsuho0XU=
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3915.0 Safari/537.36 Edg/79.0.287.3
Sec-Fetch-User: ?1
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3
Sec-Fetch-Site: none
Sec-Fetch-Mode: navigate
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
Cookie: __utmz=111872281.1557000280.1.1.utmcsr=(direct)|utmccn=(direct)|utmcmd=(none); WebStoreUser=; _ga=GA1.1.1593582071.1557000280; .AspNet.Consent=yes; weblog_user=Rick Strahl; weblog_email=rstrahl@west-wind.com; wwt=6EYWGHcnCkBkUyCL; __utma=111872281.1593582071.1557000280.1569911705.1569917482.17
```

**Comment by blowdart** [comment]
You missed one thing I asked, does IsInRole() work?
Also does it work under IIS proper?

**Comment by RickStrahl** [comment]
Sorry must have edited it ou. No it doesn’t work when the groups are missing in the claims.

**Comment by blowdart** [comment]
Weird.

Ok. @brentschmaltz you own WindowsIdentity and WindowsPrincipal. Do you have any idea what's happening here?

**Comment by RickStrahl** [comment]
Ok, I've set up a simple repo project that demonstrates the behavior I'm talking about here.

https://github.com/RickStrahl-Content/WindowsAuthWithGroupsIssue

Code is a single file [here](https://github.com/RickStrahl-Content/WindowsAuthWithGroupsIssue/blob/master/Startup.cs).

This project uses a super simple challenge and middleware endpoint to echo back what groups it sees after the login succeeds.

The two scenarios that differ are:

* Intranet Auto-logon 
* Forced Browser Dialog Logon

As mentinoed above I get very different results for the auto-login vs. a forced browser login. The Auto-login is not returning any of my custom groups. In fact it's returning quite a different set of groups.

> Note in all the shots below the User Sid is always the same - IOW, we're talking about the same user, on the same machine in the same environment, returning different results here!

**Explicit Browser Login get the full list of Groups**:

![image](https://user-images.githubusercontent.com/1374013/66519028-2a6dd580-ea82-11e9-826c-cab6615365b3.png)

This works as it should - it's retrieving the list of custom accounts I've added.

**Automatic Browser Login on Chrome, Edge, FireFox**

![image](https://user-images.githubusercontent.com/1374013/66520855-a7e71500-ea85-11e9-8bd3-ec316228e5dd.png)


## IIS Proper
Finally I also published the site (locally) and used IIS to access it. Enabled Windows Authentication on the new site, then added the site to local Intranet groups.

With Auto-logon enabled once again - no groups are returned.

![image](https://user-images.githubusercontent.com/1374013/66520902-c1885c80-ea85-11e9-8d5e-39bf874364ef.png)


To clarfify the behavior occurs in Chrome, Edgium. FireFox is Ok because it doesn't support auto-login and always forces a login. Classic Edge fails to load the page (WTF?)...

**Comment by RickStrahl** [comment]
Any update on this?

**Comment by RickStrahl** [comment]
Ok - figured it out. Turns out it's sort of Operator Error, but related to the way Windows logons are apparently handled.

The issue is that I created **new groups** in Windows and then tried to use these groups in the application. These days, I don't reboot or log out of Windows very much so even though this discussion has dragged on for a week and more, I **never logged out**.

Apparently when you do Auto-Logon, Windows picks up a cached token of the user **when the logon occurred**. IOW, it looks like it shows only the groups that were present when I logged in. When logging in explicitly it refreshes credentials completely rather than re-using the cached credential.

When I finally decided to reboot my machine, the automatic login started returning the missing groups just fine. I verified if I add additional groups after the logon, they don't show up until I either log out or reboot.

This might be worthwhile to document in relation to groups with Windows Authentication.

Many thanks to Gabriel Luci, who actually suggested this in [this StackOverflow issue](https://stackoverflow.com/a/58347463/11197) I posted related to this issue.
